content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,496 @@
+  Kubernetes API. This allows Kubernetes to fetch a list of possible permissions for your user.
+  Kubernetes then generates a personalized response to your request in real-time. This logic is
+  different from a scenario where this resource is simply stored in etcd.
+
+- Similarly, in KubeVirt's [CDI (Containerized Data Importer)](https://github.com/kubevirt/containerized-data-importer)
+  extension, which allows file uploads into a PVC from a local machine using the `virtctl` tool,
+  a special token is required before the upload process begins.
+  This token is generated by creating an UploadTokenRequest resource via the Kubernetes API. Kubernetes
+  routes (proxies) all UploadTokenRequest resource creation requests to the CDI extension API server,
+  which generates and returns the token in response.
+
+### Full control over conversion, validation, and output formatting
+
+- Your custom API server can have all the capabilities of the vanilla Kubernetes API server. The resources you create
+  in your custom API server can be validated immediately on the server side without additional webhooks
+  (with CRDs, you'd have to write webhooks for extended validation).
+  
+  You can implement transparent conversion logic between versions like `v1alpha1`, `v1beta1`, and `v1`
+  and serve multiple API versions simultaneously.
+  
+  Kubernetes allows you to serve multiple API versions for each resource type, traditionally
+  `v1alpha1`, `v1beta1` and `v1`. Only one version can be specified as the storage version.
+  All requests to other versions must be automatically converted to the version specified as storage version.
+  With CRDs, this mechanism is implemented using conversion webhooks. Whereas in an extension API server,
+  you can implement your own conversion mechanism, choose to mix up different storage versions (one
+  object might be serialized as `v1`, another as `v2`), or rely on an external backing API.
+
+- Directly implementing the Kubernetes API lets you format table output however you like and doesn't force you to follow
+  the `additionalPrinterColumns` logic in CRDs. Instead, you can write your own formatter that
+  formats the table output and custom fields in it. For example, when using `additionalPrinterColumns`,
+  you can display field values only following the JSONPath logic. In your own API server, you can generate
+  and insert values on the fly, formatting the table output as you wish.
+
+### Dynamic resource registration
+
+- The resources served by an extension api-server don't need to be pre-registered as CRDs.
+Once your api-server is registered as an `APIService`, Kubernetes starts polling it to find out what
+resource types it serves. After receiving a response, it automatically registers all available
+types for this API group. Although this isn't considered common practice, you can implement logic
+that dynamically registers the resource types you need in Kubernetes.
```

Comment:
```suggestion
  Once your extension API server is registered using an APIService, Kubernetes starts polling it to discover
  APIs and resources it can serve. After receiving a discovery response, the Kubernetes API server automatically
  registers all available types for this API group.
  Although this isn't considered common practice, you can implement logic that dynamically registers
  the resource types you need in your Kubernetes cluster.
```
?